File: src/gui/NotificationWidget.cpp
#include "gui/NotificationWidget.h"
#include "gui/Theme.h"

#include <spdlog/spdlog.h>

#include <QEasingCurve>
#include <QSizePolicy>
#include <algorithm>

namespace vtol {

// ─── helpers ─────────────────────────────────────────────────

static const char* levelColor(NotificationLevel lvl)
{
    switch (lvl) {
    case NotificationLevel::Critical: return theme::ERROR_CLR;
    case NotificationLevel::Warning:  return theme::WARNING;
    case NotificationLevel::Info:     return theme::SUCCESS;
    }
    return theme::SUCCESS;
}

static const char* levelBg(NotificationLevel lvl)
{
    switch (lvl) {
    case NotificationLevel::Critical: return theme::ERROR_BG;
    case NotificationLevel::Warning:  return theme::WARNING_BG;
    case NotificationLevel::Info:     return theme::SUCCESS_BG;
    }
    return theme::SUCCESS_BG;
}

static int levelPriority(NotificationLevel lvl)
{
    switch (lvl) {
    case NotificationLevel::Critical: return 0;
    case NotificationLevel::Warning:  return 1;
    case NotificationLevel::Info:     return 2;
    }
    return 2;
}

static const char* levelName(NotificationLevel lvl)
{
    switch (lvl) {
    case NotificationLevel::Critical: return "critical";
    case NotificationLevel::Warning:  return "warning";
    case NotificationLevel::Info:     return "info";
    }
    return "info";
}

static int defaultDurationSec(NotificationLevel lvl)
{
    switch (lvl) {
    case NotificationLevel::Info:     return 5;
    case NotificationLevel::Warning:  return 7;
    case NotificationLevel::Critical: return 10;
    }
    return 5;
}

// ═════════════════════════════════════════════════════════════
//  NotificationWidget
// ═════════════════════════════════════════════════════════════

NotificationWidget::NotificationWidget(QWidget* parent)
    : QFrame(parent)
{
    m_timer.setInterval(16);  // ~60 FPS
    connect(&m_timer, &QTimer::timeout, this, &NotificationWidget::tick);
    setupUi();
    hide();
}

void NotificationWidget::setupUi()
{
    setObjectName(QStringLiteral("notif_card"));
    setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Fixed);

    // Fade-in / fade-out
    m_opacityEffect = new QGraphicsOpacityEffect(this);
    m_opacityEffect->setOpacity(1.0);
    setGraphicsEffect(m_opacityEffect);

    m_fadeAnim = new QPropertyAnimation(m_opacityEffect, "opacity", this);
    m_fadeAnim->setDuration(200);
    m_fadeAnim->setEasingCurve(QEasingCurve::InOutSine);
    connect(m_fadeAnim, &QPropertyAnimation::finished, this, [this] {
        if (m_dismissing)
            finalizeDismiss();
    });

    // Root layout — горизонтальные поля шире для аккуратного вида
    auto* root = new QVBoxLayout(this);
    root->setContentsMargins(14, 6, 14, 6);
    root->setSpacing(0);

    // Inner card frame
    m_inner = new QFrame;
    m_inner->setObjectName(QStringLiteral("notif_inner"));
    auto* innerLay = new QVBoxLayout(m_inner);
    innerLay->setContentsMargins(0, 0, 0, 0);
    innerLay->setSpacing(0);

    // Content area (title + message + buttons)
    auto* contentWidget = new QWidget;
    contentWidget->setStyleSheet(QStringLiteral("background: transparent;"));
    auto* contentLay = new QVBoxLayout(contentWidget);
    contentLay->setContentsMargins(16, 11, 16, 10);
    contentLay->setSpacing(3);

    m_lblTitle = new QLabel;
    m_lblTitle->setWordWrap(true);
    m_lblTitle->setStyleSheet(QStringLiteral(
        "font-family: \"%1\"; font-size: 13px; font-weight: 700; "
        "border: none; background: transparent;")
        .arg(theme::FONT_FAMILY));
    contentLay->addWidget(m_lblTitle);

    m_lblMessage = new QLabel;
    m_lblMessage->setWordWrap(true);
    m_lblMessage->setStyleSheet(QStringLiteral(
        "color: %1; font-size: 12px; border: none; background: transparent;")
        .arg(theme::TEXT_SECONDARY));
    contentLay->addWidget(m_lblMessage);

    // Action buttons row
    m_btnContainer = new QWidget;
    m_btnContainer->setStyleSheet(QStringLiteral("background: transparent;"));
    m_btnLay = new QHBoxLayout(m_btnContainer);
    m_btnLay->setContentsMargins(0, 4, 0, 2);
    m_btnLay->setSpacing(6);
    m_btnLay->addStretch();
    contentLay->addWidget(m_btnContainer);
    m_btnContainer->hide();

    innerLay->addWidget(contentWidget);

    // Progress bar — inside the card, with side margins to respect rounded corners
    m_progress = new QProgressBar;
    m_progress->setTextVisible(false);
    m_progress->setFixedHeight(3);
    m_progress->setRange(0, 1000);
    m_progress->setValue(1000);
    // No border-radius on the bar itself — it's inset from the card edges
    m_progress->setStyleSheet(QStringLiteral(
        "QProgressBar { background-color: rgba(0,0,0,0); border: none; border-radius: 0px; margin: 0 1px; }"
        "QProgressBar::chunk { background-color: %1; border-radius: 0 0 7px 7px; }")
        .arg(theme::PRIMARY));

    // Wrapper gives the progress bar rounded bottom corners via clip
    auto* progressWrapper = new QWidget;
    progressWrapper->setStyleSheet(QStringLiteral("background: transparent;"));
    auto* pwLay = new QVBoxLayout(progressWrapper);
    pwLay->setContentsMargins(0, 0, 0, 0);
    pwLay->setSpacing(0);
    pwLay->addWidget(m_progress);
    innerLay->addWidget(progressWrapper);

    root->addWidget(m_inner);
}

void NotificationWidget::showNotification(const Notification& n)
{
    m_timer.stop();
    m_dismissing = false;
    m_current = n;

    int sec = (n.durationSec < 0) ? defaultDurationSec(n.level) : n.durationSec;
    m_durationMs = sec * 1000;

    auto color = levelColor(n.level);
    auto bg    = levelBg(n.level);

    // Card style
    m_inner->setStyleSheet(QStringLiteral(
        "#notif_inner { background-color: %1; border: 1px solid %2; "
        "border-left: 3px solid %3; border-radius: 8px; }")
        .arg(bg, theme::BORDER, color));

    // Title
    m_lblTitle->setText(n.title);
    m_lblTitle->setStyleSheet(QStringLiteral(
        "color: %1; font-family: \"%2\"; font-size: 13px; font-weight: 700; "
        "border: none; background: transparent;")
        .arg(color, theme::FONT_FAMILY));

    // Message
    m_lblMessage->setText(n.message);

    // Progress bar color — subtle dark background, accent chunk
    m_progress->setStyleSheet(QStringLiteral(
        "QProgressBar { background-color: rgba(0,0,0,0.2); border: none; border-radius: 0 0 7px 7px; margin: 0px; }"
        "QProgressBar::chunk { background-color: %1; border-radius: 0 0 7px 7px; }")
        .arg(color));
    m_progress->setValue(1000);

    // Action buttons
    clearButtons();
    if (!n.actions.empty()) {
        for (const auto& [label, callback] : n.actions) {
            auto* btn = new QPushButton(label);
            btn->setCursor(Qt::PointingHandCursor);
            btn->setFixedHeight(28);
            btn->setStyleSheet(QStringLiteral(
                "QPushButton { background-color: %1; color: %2; border: 1px solid %3; "
                "border-radius: 6px; padding: 0 12px; font-size: 11px; font-weight: 600; }"
                "QPushButton:hover { background-color: %4; color: %5; border-color: %6; }")
                .arg(theme::BG_INPUT, theme::TEXT_SECONDARY, theme::BORDER,
                     theme::BG_HOVER, theme::TEXT_PRIMARY, theme::BORDER_LIGHT));

            auto cb = callback;
            auto lbl = label;
            connect(btn, &QPushButton::clicked, this, [this, cb, lbl] {
                auto title = m_current ? m_current->title : QStringLiteral("?");
                SPDLOG_INFO("[Notification] action '{}' on '{}'",
                            lbl.toStdString(), title.toStdString());
                cb();
                dismiss();
            });

            m_btnLay->insertWidget(m_btnLay->count() - 1, btn); // before stretch
            m_actionButtons.push_back(btn);
        }
        m_btnContainer->show();
    } else {
        m_btnContainer->hide();
    }

    if (m_durationMs > 0) {
        m_progress->show();
        m_elapsed.start();
        m_timer.start();
    } else {
        m_progress->hide();
        SPDLOG_DEBUG("[Notification] infinite duration, timer disabled");
    }

    fadeIn();

    SPDLOG_INFO("[Notification] [{}] {} — {} ({}s)",
                levelName(n.level), n.title.toStdString(), n.message.toStdString(),
                sec);
}

void NotificationWidget::fadeIn()
{
    m_fadeAnim->stop();
    if (!isVisible()) {
        m_opacityEffect->setOpacity(0.0);
        show();
    }
    m_fadeAnim->setStartValue(m_opacityEffect->opacity());
    m_fadeAnim->setEndValue(1.0);
    m_fadeAnim->start();
}

void NotificationWidget::fadeOut()
{
    m_fadeAnim->stop();
    m_fadeAnim->setStartValue(m_opacityEffect->opacity());
    m_fadeAnim->setEndValue(0.0);
    m_fadeAnim->start();
}

void NotificationWidget::dismiss()
{
    if (!m_current || m_dismissing) return;
    m_timer.stop();
    m_dismissing = true;
    SPDLOG_DEBUG("[Notification] dismissing '{}'", m_current->title.toStdString());
    if (isVisible()) {
        fadeOut();
    } else {
        finalizeDismiss();
    }
}

void NotificationWidget::finalizeDismiss()
{
    m_dismissing = false;
    auto title = m_current ? m_current->title : QStringLiteral("?");
    m_current.reset();
    clearButtons();
    hide();
    SPDLOG_DEBUG("[Notification] dismissed '{}'", title.toStdString());
    emit notificationClosed();
}

void NotificationWidget::tick()
{
    int elapsedMs = static_cast<int>(m_elapsed.elapsed());
    int remaining = (std::max)(0, m_durationMs - elapsedMs);
    m_progress->setValue(m_durationMs > 0 ? (remaining * 1000 / m_durationMs) : 0);
    if (remaining <= 0) {
        SPDLOG_DEBUG("[Notification] auto-closing '{}'",
                     m_current ? m_current->title.toStdString() : "?");
        dismiss();
    }
}

void NotificationWidget::clearButtons()
{
    for (auto* btn : m_actionButtons) {
        m_btnLay->removeWidget(btn);
        btn->deleteLater();
    }
    m_actionButtons.clear();
}

std::optional<NotificationLevel> NotificationWidget::currentLevel() const
{
    if (!m_current) return std::nullopt;
    return m_current->level;
}

bool NotificationWidget::currentHasActions() const
{
    return m_current.has_value() && !m_current->actions.empty();
}

void NotificationWidget::curtailToPercent(int pct)
{
    if (!m_current || m_durationMs == 0) return;  // infinite — не трогать
    int elapsedMs = static_cast<int>(m_elapsed.elapsed());
    int remaining = m_durationMs - elapsedMs;
    int cap = elapsedMs + remaining * pct / 100;
    if (cap < m_durationMs) {
        m_durationMs = cap;
        SPDLOG_DEBUG("[Notification] curtailed to {}% of remaining (~{}ms left)",
                     pct, m_durationMs - elapsedMs);
    }
}

// ═════════════════════════════════════════════════════════════
//  NotificationManager
// ═════════════════════════════════════════════════════════════

NotificationManager::NotificationManager(NotificationWidget* widget, QObject* parent)
    : QObject(parent), m_widget(widget)
{
    connect(widget, &NotificationWidget::notificationClosed,
            this, &NotificationManager::showNext);
}

void NotificationManager::setDefaultDurations(int infoSec, int warningSec, int criticalSec)
{
    m_infoDurSec     = infoSec;
    m_warningDurSec  = warningSec;
    m_criticalDurSec = criticalSec;
    SPDLOG_DEBUG("[NotificationManager] default durations: info={}s, warning={}s, critical={}s",
                 infoSec, warningSec, criticalSec);
}

void NotificationManager::setCurtailPercent(int pct)
{
    m_curtailPct = pct;
    SPDLOG_DEBUG("[NotificationManager] curtail percent: {}%", pct);
}

int NotificationManager::resolveDuration(const Notification& n) const
{
    if (n.durationSec >= 0) return n.durationSec;
    switch (n.level) {
    case NotificationLevel::Info:     return m_infoDurSec;
    case NotificationLevel::Warning:  return m_warningDurSec;
    case NotificationLevel::Critical: return m_criticalDurSec;
    }
    return m_infoDurSec;
}

void NotificationManager::push(const Notification& n)
{
    Notification resolved = n;
    resolved.durationSec = resolveDuration(n);

    insertByPriority(resolved);
    if (!resolved.tag.isEmpty())
        m_tags[resolved.tag.toStdString()] = 1;

    // --- interrupt check ---
    if (m_widget->isShowing() && !m_widget->currentHasActions()) {
        auto curLvl = m_widget->currentLevel();
        bool incomingHigher   = curLvl && levelPriority(resolved.level) < levelPriority(*curLvl);
        bool incomingDecision = !resolved.actions.empty();
        if (incomingHigher || incomingDecision) {
            SPDLOG_INFO("[NotificationQueue] interrupting current (curtail to {}% of remaining), incoming '{}' [{}]",
                        m_curtailPct, resolved.title.toStdString(), levelName(resolved.level));
            m_widget->curtailToPercent(m_curtailPct);
        }
    }
    // --- end interrupt check ---

    SPDLOG_DEBUG("[NotificationQueue] push '{}' [{}], duration={}s, queue_size={}",
                 resolved.title.toStdString(), levelName(resolved.level),
                 resolved.durationSec, m_queue.size());

    if (!m_widget->isShowing())
        showNext();
}

void NotificationManager::pushOrReplace(Notification n, const QString& tag)
{
    n.tag = tag;
    auto key = tag.toStdString();
    if (m_tags.count(key)) {
        auto it = std::find_if(m_queue.begin(), m_queue.end(),
            [&tag](const Notification& x) { return x.tag == tag; });
        if (it != m_queue.end())
            m_queue.erase(it);
    }
    push(n);
}

void NotificationManager::clear()
{
    m_queue.clear();
    m_tags.clear();
    m_widget->dismiss();
    SPDLOG_INFO("[NotificationQueue] cleared");
}

void NotificationManager::insertByPriority(const Notification& n)
{
    int prio = levelPriority(n.level);
    for (auto it = m_queue.begin(); it != m_queue.end(); ++it) {
        if (levelPriority(it->level) > prio) {
            m_queue.insert(it, n);
            return;
        }
    }
    m_queue.push_back(n);
}

void NotificationManager::showNext()
{
    if (!m_queue.empty()) {
        auto notif = std::move(m_queue.front());
        m_queue.pop_front();
        if (!notif.tag.isEmpty())
            m_tags.erase(notif.tag.toStdString());
        SPDLOG_DEBUG("[NotificationQueue] showing next, remaining={}",
                     m_queue.size());
        m_widget->showNotification(notif);
    } else {
        m_widget->hide();
    }
}

} // namespace vtol

File: src/gui/notification_widget.py
"""
Notification widget — toast-like card with title, message, action buttons,
and a countdown progress bar.  Designed for the right sidebar panel.
"""
import logging
from collections import deque
from dataclasses import dataclass, field
from enum import Enum
from typing import Callable, List, Optional, Tuple

from PyQt5.QtCore import Qt, QTimer, pyqtSignal, QObject
from PyQt5.QtWidgets import (
    QFrame, QVBoxLayout, QHBoxLayout, QLabel, QPushButton,
    QProgressBar, QWidget, QSizePolicy,
)

from src.gui.theme import Colors, Fonts

logger = logging.getLogger(__name__)


# ─── Data types ───────────────────────────────────────────────

class NotificationLevel(Enum):
    INFO = "info"
    WARNING = "warning"
    CRITICAL = "critical"


_LEVEL_COLORS = {
    NotificationLevel.INFO: Colors.SUCCESS,
    NotificationLevel.WARNING: Colors.WARNING,
    NotificationLevel.CRITICAL: Colors.ERROR,
}

_LEVEL_BG = {
    NotificationLevel.INFO: Colors.SUCCESS_BG,
    NotificationLevel.WARNING: Colors.WARNING_BG,
    NotificationLevel.CRITICAL: Colors.ERROR_BG,
}


@dataclass
class Notification:
    level: NotificationLevel
    title: str
    message: str
    duration: int = 60          # seconds
    actions: List[Tuple[str, Callable]] = field(default_factory=list)
    tag: str = ""               # for deduplication


# ─── Widget ───────────────────────────────────────────────────

class NotificationWidget(QFrame):
    """Single notification card shown in the sidebar."""

    notification_closed = pyqtSignal()

    _TICK_MS = 100

    def __init__(self, parent: QWidget = None):
        super().__init__(parent)
        self._current: Optional[Notification] = None
        self._elapsed_ms: int = 0
        self._duration_ms: int = 0

        self._timer = QTimer(self)
        self._timer.setInterval(self._TICK_MS)
        self._timer.timeout.connect(self._tick)

        self._action_buttons: List[QPushButton] = []

        self._setup_ui()
        self.hide()

    # ── UI ──

    def _setup_ui(self):
        self.setObjectName("notif_card")
        self.setSizePolicy(QSizePolicy.Expanding, QSizePolicy.Fixed)

        root = QVBoxLayout(self)
        root.setContentsMargins(0, 0, 0, 0)
        root.setSpacing(0)

        # Inner frame with left-border accent
        self._inner = QFrame()
        self._inner.setObjectName("notif_inner")
        inner_lay = QVBoxLayout(self._inner)
        inner_lay.setContentsMargins(12, 10, 12, 6)
        inner_lay.setSpacing(4)

        # Title
        self._lbl_title = QLabel()
        self._lbl_title.setWordWrap(True)
        self._lbl_title.setStyleSheet(f"""
            font-family: "{Fonts.FAMILY}";
            font-size: 13px;
            font-weight: 700;
            border: none;
            background: transparent;
        """)
        inner_lay.addWidget(self._lbl_title)

        # Message
        self._lbl_message = QLabel()
        self._lbl_message.setWordWrap(True)
        self._lbl_message.setStyleSheet(f"""
            color: {Colors.TEXT_SECONDARY};
            font-size: 12px;
            border: none;
            background: transparent;
        """)
        inner_lay.addWidget(self._lbl_message)

        # Action buttons row
        self._btn_container = QWidget()
        self._btn_container.setStyleSheet("background: transparent;")
        self._btn_lay = QHBoxLayout(self._btn_container)
        self._btn_lay.setContentsMargins(0, 4, 0, 2)
        self._btn_lay.setSpacing(6)
        self._btn_lay.addStretch()
        inner_lay.addWidget(self._btn_container)
        self._btn_container.hide()

        root.addWidget(self._inner)

        # Progress bar (bottom edge)
        self._progress = QProgressBar()
        self._progress.setTextVisible(False)
        self._progress.setFixedHeight(3)
        self._progress.setRange(0, 1000)
        self._progress.setValue(1000)
        self._progress.setStyleSheet(f"""
            QProgressBar {{
                background-color: {Colors.BG_INPUT};
                border: none;
                border-radius: 0px;
            }}
            QProgressBar::chunk {{
                background-color: {Colors.PRIMARY};
                border-radius: 0px;
            }}
        """)
        root.addWidget(self._progress)

    # ── Public API ──

    def show_notification(self, notification: Notification):
        self._stop()
        self._current = notification
        self._elapsed_ms = 0
        self._duration_ms = notification.duration * 1000

        color = _LEVEL_COLORS[notification.level]
        bg = _LEVEL_BG[notification.level]

        # Card style
        self._inner.setStyleSheet(f"""
            #notif_inner {{
                background-color: {bg};
                border: 1px solid {Colors.BORDER};
                border-left: 3px solid {color};
                border-radius: 8px;
            }}
        """)

        # Title color
        self._lbl_title.setText(notification.title)
        self._lbl_title.setStyleSheet(f"""
            color: {color};
            font-family: "{Fonts.FAMILY}";
            font-size: 13px;
            font-weight: 700;
            border: none;
            background: transparent;
        """)

        # Message
        self._lbl_message.setText(notification.message)

        # Progress bar color
        self._progress.setStyleSheet(f"""
            QProgressBar {{
                background-color: {Colors.BG_INPUT};
                border: none;
                border-radius: 0px;
            }}
            QProgressBar::chunk {{
                background-color: {color};
                border-radius: 0px;
            }}
        """)
        self._progress.setValue(1000)

        # Action buttons
        self._clear_buttons()
        if notification.actions:
            for label, callback in notification.actions:
                btn = QPushButton(label)
                btn.setCursor(Qt.PointingHandCursor)
                btn.setFixedHeight(28)
                btn.setStyleSheet(f"""
                    QPushButton {{
                        background-color: {Colors.BG_INPUT};
                        color: {Colors.TEXT_SECONDARY};
                        border: 1px solid {Colors.BORDER};
                        border-radius: 6px;
                        padding: 0 12px;
                        font-size: 11px;
                        font-weight: 600;
                    }}
                    QPushButton:hover {{
                        background-color: {Colors.BG_HOVER};
                        color: {Colors.TEXT_PRIMARY};
                        border-color: {Colors.BORDER_LIGHT};
                    }}
                """)
                # capture label/callback in closure
                btn.clicked.connect(self._make_action_handler(label, callback))
                self._btn_lay.insertWidget(self._btn_lay.count() - 1, btn)
                self._action_buttons.append(btn)
            self._btn_container.show()
        else:
            self._btn_container.hide()

        self.show()
        self._timer.start()

        logger.info("NOTIFICATION: [%s] %s — %s",
                     notification.level.value, notification.title, notification.message)

    def dismiss(self):
        if self._current:
            title = self._current.title
            self._stop()
            self._current = None
            self.hide()
            logger.debug("NOTIFICATION: dismissed '%s'", title)
            self.notification_closed.emit()

    def is_showing(self) -> bool:
        return self._current is not None

    # ── Private ──

    def _tick(self):
        self._elapsed_ms += self._TICK_MS
        remaining = max(0, self._duration_ms - self._elapsed_ms)
        self._progress.setValue(int(remaining * 1000 / self._duration_ms) if self._duration_ms else 0)
        if remaining <= 0:
            title = self._current.title if self._current else "?"
            self._stop()
            self._current = None
            self.hide()
            logger.debug("NOTIFICATION: auto-closed '%s'", title)
            self.notification_closed.emit()

    def _stop(self):
        self._timer.stop()
        self._clear_buttons()

    def _clear_buttons(self):
        for btn in self._action_buttons:
            self._btn_lay.removeWidget(btn)
            btn.deleteLater()
        self._action_buttons.clear()

    def _make_action_handler(self, label: str, callback: Callable):
        def handler():
            title = self._current.title if self._current else "?"
            logger.info("NOTIFICATION: action '%s' on '%s'", label, title)
            callback()
            self.dismiss()
        return handler


# ─── Manager (queue) ──────────────────────────────────────────

class NotificationManager(QObject):
    """Priority queue that feeds notifications one-by-one into the widget."""

    _PRIORITY = {
        NotificationLevel.CRITICAL: 0,
        NotificationLevel.WARNING: 1,
        NotificationLevel.INFO: 2,
    }

    def __init__(self, widget: NotificationWidget, parent: QObject = None):
        super().__init__(parent)
        self._widget = widget
        self._queue: deque[Notification] = deque()
        self._tags: dict[str, Notification] = {}
        widget.notification_closed.connect(self._show_next)

    def push(self, notification: Notification):
        """Add a notification. Shows immediately if nothing is displayed."""
        self._insert_by_priority(notification)
        if notification.tag:
            self._tags[notification.tag] = notification
        logger.debug("NOTIFICATION_QUEUE: push '%s' [%s], queue_size=%d",
                      notification.title, notification.level.value, len(self._queue))
        if not self._widget.is_showing():
            self._show_next()

    def push_or_replace(self, notification: Notification, tag: str):
        """Replace an existing notification with the same tag, or add new."""
        notification.tag = tag
        if tag in self._tags:
            old = self._tags[tag]
            # Remove old from queue if still queued
            try:
                self._queue.remove(old)
            except ValueError:
                pass
            # If old is currently showing — dismiss and show replacement
            if (self._widget.is_showing()
                    and self._widget._current is old):
                self._widget.dismiss()
        self.push(notification)

    def clear(self):
        """Clear the queue and dismiss current notification."""
        self._queue.clear()
        self._tags.clear()
        self._widget.dismiss()
        logger.info("NOTIFICATION_QUEUE: cleared")

    # ── Private ──

    def _insert_by_priority(self, notification: Notification):
        prio = self._PRIORITY[notification.level]
        # Find insertion point — after all items with same or higher priority
        for i, existing in enumerate(self._queue):
            if self._PRIORITY[existing.level] > prio:
                self._queue.insert(i, notification)
                return
        self._queue.append(notification)

    def _show_next(self):
        if self._queue:
            notif = self._queue.popleft()
            if notif.tag and notif.tag in self._tags:
                del self._tags[notif.tag]
            logger.debug("NOTIFICATION_QUEUE: showing next, remaining=%d", len(self._queue))
            self._widget.show_notification(notif)
        else:
            self._widget.hide()
